File: multithreading.java
import java.util.*;


public class  multithreading {
    static final int N = 1000000;
    static final int NUM_THREADS = 4;
    static final List<Integer> primes = Collections.synchronizedList(new ArrayList<>());

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();

        // Get memory usage before computation
        Runtime runtime = Runtime.getRuntime();
        runtime.gc(); // suggest garbage collection
        long startMemory = runtime.totalMemory() - runtime.freeMemory();

        int chunk = N / NUM_THREADS;
        Thread[] threads = new Thread[NUM_THREADS];

        for (int i = 0; i < NUM_THREADS; i++) {
            int start = i * chunk;
            int end = (i == NUM_THREADS - 1) ? N : (i + 1) * chunk;
            threads[i] = new Thread(new PrimeWorker(start, end));
            threads[i].start();
        }

        for (int i = 0; i < NUM_THREADS; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        long endTime = System.currentTimeMillis();
        long endMemory = runtime.totalMemory() - runtime.freeMemory();

        // Output
        System.out.println("Total primes found: " + primes.size());
        System.out.printf("⏱ Time taken        : %.2f seconds%n", (endTime - startTime) / 1000.0);
        System.out.printf(" Peak memory used : %.2f MB%n", (endMemory - startMemory) / (1024.0 * 1024.0));
    }

    static class PrimeWorker implements Runnable {
        int start, end;

        PrimeWorker(int start, int end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public void run() {
            List<Integer> localPrimes = new ArrayList<>();
            for (int i = start; i < end; i++) {
                if (isPrime(i)) {
                    localPrimes.add(i);
                }
            }
            primes.addAll(localPrimes);
        }

        boolean isPrime(int n) {
            if (n <= 1) return false;
            if (n == 2) return true;
            if (n % 2 == 0) return false;
            int sqrt = (int) Math.sqrt(n);
            for (int i = 3; i <= sqrt; i += 2) {
                if (n % i == 0) return false;
            }
            return true;
        }
    }
}

File: multithreading.py
#multithreading
import threading
import time
import tracemalloc
import math

# Parameters
N = 1000000
NUM_THREADS = 4
primes = []

# Thread-safe lock for appending results
lock = threading.Lock()

# Prime check function
def is_prime(n):
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    sqrt_n = int(math.sqrt(n)) + 1
    for i in range(3, sqrt_n, 2):
        if n % i == 0:
            return False
    return True

# Worker function
def find_primes(start, end):
    local_primes = []
    for number in range(start, end):
        if is_prime(number):
            local_primes.append(number)
    with lock:
        primes.extend(local_primes)

# Main
if __name__ == "__main__":
    tracemalloc.start()
    start_time = time.time()

    threads = []
    chunk = N // NUM_THREADS

    for i in range(NUM_THREADS):
        start = i * chunk
        end = N if i == NUM_THREADS - 1 else (i + 1) * chunk
        t = threading.Thread(target=find_primes, args=(start, end))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    end_time = time.time()
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print(f"Total primes found: {len(primes)}")
    print(f"Time taken        : {end_time - start_time:.2f} seconds")
    print(f"Peak memory used : {peak / (1024 * 1024):.2f} MB")